src/Features/Core/Portable/Snippets/SnippetProviders/AbstractConstructorSnippetProvider.cs

```diff
@@ -32,12 +32,25 @@ protected override async Task<TextChange> GenerateSnippetTextChangeAsync(Documen
             var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
             var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
             var nodeAtPosition = root.FindNode(TextSpan.FromBounds(position, position));
-            var containingType = nodeAtPosition.FirstAncestorOrSelf<SyntaxNode>(syntaxFacts.IsTypeDeclaration);
+
+            // Skip inner class in nested class if position is out of it.
+            //     For example "class Outer{\r\n ctor@ \r\n class Inner {} }"
+            // Accept even if "static class" (CS0710).
+            var containingType = nodeAtPosition.FirstAncestorOrSelf<SyntaxNode>((node) =>
+            {
+                return syntaxFacts.IsTypeDeclaration(node)
+                    && IsConstructableTypeDeclaration(syntaxFacts, node)
+                    && node.Span.Contains(position);
+            });
+
             Contract.ThrowIfNull(containingType);
             var constructorDeclaration = generator.ConstructorDeclaration(
                 containingTypeName: syntaxFacts.GetIdentifierOfTypeDeclaration(containingType).ToString(),
                 accessibility: Accessibility.Public);
             return new TextChange(TextSpan.FromBounds(position, position), constructorDeclaration.NormalizeWhitespace().ToFullString());
         }
+
+        // Move this to ISyntaxFacts if approved.
+        protected abstract bool IsConstructableTypeDeclaration(ISyntaxFacts syntaxFacts, SyntaxNode node);
```

Comment:
Is this necessary? There's already a check in `CSharpConstructorSnippetProvider.IsValidSnippetLocationAsync`

Reply:
@genlu  Isn't IsValidSnippetLocationAsync  check that the constructor can be placed where 'ctor' is entered?
I think that function is not suitable to get the target class name.

Is it acceptable that the decisions required within AbstractConstructorSnippetProvider depend on the processing implemented in the classes that inherited AbstractConstructorSnippetProvider?
I can't think of any other way to determine the constructable type from within an abstract class.

Reply:
You are right, IsValidSnippetLocationAsync](https://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/Snippets/SnippetProviders/AbstractSnippetProvider.cs,73) is used to decide whether a snippet should placed at the location. So if the check in IsValidSnippetLocationAsync fails, we won't even show the snippet thus it will never get to the point where code change is being calculated.

So to answer your question: yes, it's not only acceptable, it's actually supposed to work this way by design.